### PR TITLE
Re-Land "Refactor StrokeAlign to allow double values." (#108339)

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This re-lands #108339 with some updates that keep the rendering the same as it was before so that they don't break golden tests. There are still some acceptable very small (one grey level) rounding differences in `BoxBorder` rendering, but everything else is closer to where it should have been, especially the corners on zero-width rounded rectangles, and zero-radius corners on `ContinuousRectangleBorder`s.

I also renamed the sample, and linked it to the `BorderSide.strokeAlign` member docs.

The first commit is the original re-land, and the subsequent commits are my changes.

Related Issue: https://github.com/flutter/flutter/issues/106362

cc @bernaferrari 

internal reference: b/243217872
G3Fix: cl/469882656

### CODE REVIEW

File: packages/flutter/lib/src/painting/rounded_rectangle_border.dart [hunk 1]
```diff
@@ -126,12 +126,17 @@ class RoundedRectangleBorder extends OutlinedBorder {
...
+          canvas.drawRRect(borderRadius.resolve(textDirection).toRRect(rect), side.toPaint());
+        } else {
+          final Paint paint = Paint()
+            ..color = side.color;
+          final RRect borderRect = borderRadius.resolve(textDirection).toRRect(rect);
+          final RRect inner = borderRect.deflate(side.strokeInset);
+          final RRect outer = borderRect.inflate(side.strokeOutset);
+          canvas.drawDRRect(outer, inner, paint);
+        }
+        break;
```
Comment (bernaferrari) [inline packages/flutter/lib/src/painting/rounded_rectangle_border.dart:139]: Is `canvas.drawRRect` better than `break;`? I'm curious if there was a difference.

Your `break;` in 139 is 'unecessary' (if you add, you might want to add in 130, after drawRRect, and on the other borders. But `break` = nothing so it doesn't matter.

**Comment by bernaferrari** [comment]
Thanks for your persistence on this!

File: packages/flutter/lib/src/painting/rounded_rectangle_border.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/painting/rounded_rectangle_border.dart:139]: I realize that it doesn't do anything, but it provides consistency with the other case statement, and if a new case is added, the person adding it won't need to add a `break`.

**Comment by bernaferrari** [comment]
> I also added an additional sample to the BorderSide.strokeAlign member.

Isn't this the previous sample that we made and got merged/reverted/re-merged/re-reverted? I'm kind of confused. But everything is ok.

**Comment by gspencergoog** [comment]
> Isn't this the previous sample that we made and got merged/reverted/re-merged/re-reverted? I'm kind of confused. But everything is ok.

Yeah, could be. :-). I just noticed it wasn't linked in the docs, and that the files were sitting on my disk, so I added them.

File: packages/flutter/lib/src/painting/rounded_rectangle_border.dart (hunk 1 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/rounded_rectangle_border.dart:139]: Oh, sure. Feel free to add to the other borders if you feel it is relevant, I think only the Continous or Beveled have it. The other ones don't.

File: packages/flutter/lib/src/painting/rounded_rectangle_border.dart [hunk 2]
```diff
@@ -342,18 +293,7 @@ class _RoundedRectangleToCircleBorder extends OutlinedBorder {
...
-        adjustedRect = borderRect.deflate(side.width);
-        break;
-      case StrokeAlign.center:
-        adjustedRect = borderRect.deflate(side.width / 2);
-        break;
-      case StrokeAlign.outside:
-        adjustedRect = borderRect;
-        break;
-    }
+    final RRect adjustedRect = borderRect.deflate(ui.lerpDouble(side.width, 0, side.strokeAlign)!);
```
Comment (bernaferrari) [inline packages/flutter/lib/src/painting/rounded_rectangle_border.dart:296]: I think you are using an older version. The one that was merged didn't have this. We replace the lerp with strokeInset.
 
image

Comment (gspencergoog) [reply packages/flutter/lib/src/painting/rounded_rectangle_border.dart:296]: Wait... I'm not seeing the same thing you are.

In both the [originally landed PR](https://github.com/flutter/flutter/pull/108339/files#diff-87b32418abb85e9d4e5f0c1e3b1806e0ea53d9562ac8ed866a2f76d32928251eR279), and [here](https://github.com/flutter/flutter/pull/110234/files/52cca658583860723d94f36809872e98851e7e13#diff-87b32418abb85e9d4e5f0c1e3b1806e0ea53d9562ac8ed866a2f76d32928251eR296), the line is:

```
final RRect adjustedRect = borderRect.deflate(ui.lerpDouble(side.width, 0, side.strokeAlign)!);
```

You're looking at line 100 on the right, not line 296.

The real question is: should that be changed?  It seems to work as-is.

Comment (gspencergoog) [reply packages/flutter/lib/src/painting/rounded_rectangle_border.dart:296]: You're right, though, that the example is the same as the original, I just linked it into the docs, and renamed it to match the naming convention.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/painting/rounded_rectangle_border.dart (hunk 2 above)
Comment (bernaferrari) [reply packages/flutter/lib/src/painting/rounded_rectangle_border.dart:296]: Ok, forgive me. I got confused and looked in the wrong place. It was late at night 😅.

Comment (gspencergoog) [reply packages/flutter/lib/src/painting/rounded_rectangle_border.dart:296]: No worries.

**Comment by bernaferrari** [comment]
I just double-checked everything, everything looks good to me.

File: packages/flutter/lib/src/painting/borders.dart [hunk 3]
```diff
@@ -21,27 +21,6 @@ enum BorderStyle {
...
-/// The relative position of the stroke on a [BorderSide] in a [Border] or [OutlinedBorder].
-/// When set to [inside], the stroke is drawn completely inside the widget.
-/// For [center] and [outside], a property such as [Container.clipBehavior]
-/// can be used in an outside widget to clip it.
-/// If [Container.decoration] has a border, the container may incorporate
-/// [BorderSide.width] as additional padding:
-/// - [inside] provides padding with full [BorderSide.width].
-/// - [center] provides padding with half [BorderSide.width].
-/// - [outside] provides zero padding, as stroke is drawn entirely outside.
-enum StrokeAlign {
```
Comment (werainkhatri) [inline packages/flutter/lib/src/painting/borders.dart:33]: why is removing this not a breaking change? the code I wrote a few months ago became red w/o any pointers on why or how to fix it. i had to go into the flutter code, see the last changing commit and go through the PR to understand the reason and cause of the breakage. am i missing something?

Comment (bernaferrari) [reply packages/flutter/lib/src/painting/borders.dart:33]: AFAIK it is complicated, there is no easy way to keep that without making a weird class that acts as a shortcut for doubles, even though I tried, it was weird and non-standard, so BorderSide.strokeAlignInside became the default (you can read the discussion in the original PR). The "part 1" and "part 2" commits were like 2 months apart and strokeAlign was still really new when it was modified, but due to how Flutter stable works, they ended up being like 6 months apart.

Comment (werainkhatri) [reply packages/flutter/lib/src/painting/borders.dart:33]: hmm that makes sense. if only dart supported union types natively.

Comment (bernaferrari) [reply packages/flutter/lib/src/painting/borders.dart:33]: Me too, but we can only hope they prioritize that for 2024.

**Pull request merged**